_summaries/he2019momentum_2.md

```diff
@@ -3,17 +3,100 @@ layout: summary
+
+- Contrastive Learning is a loss that can be used in a pretext task, and is usually focused around comparing the similarities of two representations (an example is the bullet point above)
+
+## How does contrastive learning work
+
+- Given some input $$q$$ and a list of keys $$k_i$$ we want to have a low value for when $$q$$ and $$k_+$$ have similar representation and have a high value when they are different.
+
+- One way to measure the similarity between queries and keys is by the dot product.
+
+- We can then measure the basic loss of a contrastive algorithm by:
+
+$$L_q = - log \frac{exp(q \cdot k_+ / t)}{\sum^K_{t=0}exp(q \cdot k_i / t)}$$
+
+- where $$q$$ is the current image query, $$k_+$$ is the correct key that corresponds to the query, $$k_i$$ are negative samples, and $$t$$ is a temperature hyperparameter.
+
+## Why is a Queue and Momentum useful for Encoding keys in Contrastive Learning (technical implementation)
+
+![MoCo Model](./he2019momentum_2a.jpeg)
+
+(queue is not shown in this diagram)
+
+- You have to have sufficient numbers of negative samples to get good representations
+
+- At the same time, you don't want to compare keys that were encoded by an outdated previous iteration of the network 
+
+- A queue allows you to accumulate negative samples, but age out sufficiently old samples that were generated by an older version of the network
+
+- Momentum allows you to use the parameter updates from the query encoder on the key encoder but still produce consistant encodings for the keys.
+
+$$\theta_k \leftarrow m\theta_k + (1-m)\theta_q$$
+
+- where $$\theta_k$$ is the Key Encoder network and $$theta_q$$ is the Query encoding network, $$m$$ is how much momentum the query encoders backprop values are allowed to influence the key encoder.
+
+- MoCo introduces both of these ideas in their implementation.
+
+## How does it perform
+
+- MoCo was trained on Image Net 1k (authors call it 1m because each image is a "class" in their pretext task) as well as Instagram 1 Billion (940million public images)
+
+- MoCo was evaluated agianst other pretraining models on the same pretext task.
+
+- MoCo was evaluated against supervised pretraining and randomly initialized models for downstream tasks 
+    - Because MoCo is fine-tuned on downstream tasks, the experiment was setup, so the fine-tuning was as long as the supervised/random initialized training so that direct comparisons could take place.
+
+
+
+- Extremely well!  It outperforms all other unsupervised pretext models.
+
+<p align="center">
+  <img width="1200" alt="Pretext Tasks" src="./he2019momentum_2d.png">
+</p>
+
+- MoCo also outperforms every model on nearly every down stream task (note the one failure being semantic segmentations on VOC)
+
+<p align="center">
+  <img width="1200" alt="Downstream Tasks" src="./he2019momentum_2c.png">
+</p>
+
```

Comment:
I think there are more results and tables you can look to include, such as on VOC object detection.

Reply:
Oh right, I added that result table.